### ISSUE TITLE
Actor type don't failover when app crashes but sidecar remains healthy

### ISSUE THREAD

**Issue by artursouza** [dapr/dapr#1070]
## In what area(s)?

> /area runtime
> /area operator
/area placement
> /area docs
> /area test-and-release

## What version of Dapr?

0.4.x

## Expected Behavior

Placement service should be aware of the health of the application for a given actor type, not only the sidecar instance. If the app crashes but another app is up and registered for the same actortype, then calls to actor instance (new or existing Id) fail. In other words, the sidecar should report the health of the app too, so placement service can trigger rebalancing.

Client should successfully invoke actor methods:
```
== APP == Actor 764eddc2-9155-426e-b406-0da1eb48cefa got a reply: 2020-02-19 13:37:44.801

== APP == Actor 764eddc2-9155-426e-b406-0da1eb48cefa got a reply: 2020-02-19 13:37:45.018

== APP == Actor f955f205-0955-4f96-a021-6c6ce0e21421 got a reply: 2020-02-19 13:37:45.075

== APP == Actor 55f0f686-f125-4fe6-85d4-2a7bd37fb44e got a reply: 2020-02-19 13:37:45.242

== APP == Actor f955f205-0955-4f96-a021-6c6ce0e21421 got a reply: 2020-02-19 13:37:45.884

== APP == Actor 764eddc2-9155-426e-b406-0da1eb48cefa got a reply: 2020-02-19 13:37:45.940

== APP == Actor 55f0f686-f125-4fe6-85d4-2a7bd37fb44e got a reply: 2020-02-19 13:37:46.163

== APP == Actor f955f205-0955-4f96-a021-6c6ce0e21421 got a reply: 2020-02-19 13:37:46.422

== APP == Actor 764eddc2-9155-426e-b406-0da1eb48cefa got a reply: 2020-02-19 13:37:46.760

== APP == Actor f955f205-0955-4f96-a021-6c6ce0e21421 got a reply: 2020-02-19 13:37:46.788
```

## Actual Behavior

Error in client:
```
== APP == io.dapr.exceptions.DaprException: ERR_ACTOR_INVOKE_METHOD: rpc error: code = Unknown desc = error activating actor type DemoActor with id 60dad782-a802-4b39-acff-0d5487471756: dial tcp4 127.0.0.1:3000: connectex: No connection could be made because the target machine actively refused it.

== APP ==       at io.dapr.client.DaprHttp.lambda$invokeApi$3(DaprHttp.java:201)

== APP ==       at reactor.core.publisher.MonoCallable.subscribe(MonoCallable.java:56)

== APP ==       at reactor.core.publisher.Mono.subscribe(Mono.java:4105)

== APP ==       at reactor.core.publisher.Mono.block(Mono.java:1662)

== APP ==       at io.dapr.examples.actors.http.DemoActorClient.callActorForever(DemoActorClient.java:61)

== APP ==       at io.dapr.examples.actors.http.DemoActorClient.lambda$main$0(DemoActorClient.java:42)

== APP ==       at java.base/java.lang.Thread.run(Thread.java:834)

== APP ==       Suppressed: java.lang.Exception: #block terminated with an error

== APP ==               at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:93)

== APP ==               at reactor.core.publisher.Mono.block(Mono.java:1663)

== APP ==               ... 3 more

```

## Steps to Reproduce the Problem

Checkout Java SDK, have Maven and Java SDK installed.

```
mvn clean install
dapr run --app-id demoactorservice --app-port 3000 -- java -jar examples/target/dapr-java-sdk-examples-exec.jar io.dapr.examples.actors.http.DemoActorService -p 3000
```

Wait for service to start. Open a new terminal (terminal #2), then:

```
dapr run --app-id demoactorclient -- java -jar examples/target/dapr-java-sdk-examples-exec.jar io.dapr.examples.actors.http.DemoActorClient
```

Open a third terminal, then:
```
pskill java.exe
dapr run --app-id demoactorservice2 --app-port 3001 -- java -jar examples/target/dapr-java-sdk-examples-exec.jar io.dapr.examples.actors.http.DemoActorService -p 3001
```

Go back to terminal #2 and rerun:
```
dapr run --app-id demoactorclient -- java -jar examples/target/dapr-java-sdk-examples-exec.jar io.dapr.examples.actors.http.DemoActorClient
```

Now, see errors:
```
== APP == io.dapr.exceptions.DaprException: ERR_ACTOR_INVOKE_METHOD: rpc error: code = Unknown desc = error activating actor type DemoActor with id 60dad782-a802-4b39-acff-0d5487471756: dial tcp4 127.0.0.1:3000: connectex: No connection could be made because the target machine actively refused it.

== APP ==       at io.dapr.client.DaprHttp.lambda$invokeApi$3(DaprHttp.java:201)

== APP ==       at reactor.core.publisher.MonoCallable.subscribe(MonoCallable.java:56)

== APP ==       at reactor.core.publisher.Mono.subscribe(Mono.java:4105)

== APP ==       at reactor.core.publisher.Mono.block(Mono.java:1662)

== APP ==       at io.dapr.examples.actors.http.DemoActorClient.callActorForever(DemoActorClient.java:61)

== APP ==       at io.dapr.examples.actors.http.DemoActorClient.lambda$main$0(DemoActorClient.java:42)

== APP ==       at java.base/java.lang.Thread.run(Thread.java:834)

== APP ==       Suppressed: java.lang.Exception: #block terminated with an error

== APP ==               at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:93)

== APP ==               at reactor.core.publisher.Mono.block(Mono.java:1663)

== APP ==               ... 3 more

```

**Comment by yaron2** [comment]
This only happens locally, not in a clustered environment like k8s. This is because when running locally, the two processes register as the same node, and no rebalancing happens.

I guess we can close this issue.

**Comment by artursouza** [comment]
@yaron2 Like we discussed (updated issue title and description to reflect this), the issue here is that the sidecar health should not be enough for the placement service to determine rebalancing. The sidecar should also send the health status of the app, so placement service can rebalance an actor type if app is unhealthy (continuing restarting in case of K8s).

**Comment by yaron2** [comment]
@amanbha is this a dupe of something you opened in the past? My long-term memory is coming into play here..

**Comment by amanbha** [comment]
> @amanbha is this a dupe of something you opened in the past? My long-term memory is coming into play here..

@yaron2 It seems different, I think the issue which you are referring to is this one https://github.com/dapr/dapr/issues/1038

**Comment by yaron2** [comment]
> > @amanbha is this a dupe of something you opened in the past? My long-term memory is coming into play here..
> 
> @yaron2 It seems different, I think the issue which you are referring to is this one #1038

Yeah that's the one. The common theme is we should get the runtime to detect app restarts. This is useful for metrics too.

**Comment by amanbha** [comment]
About this issue: This is a broader question than placements, should Dapr ask user apps for health probes to handle errors for pub-sub/invocation etc.
I think, if the user service has crashed, then user has to bring it up based on monitoring, while the instance is down, the calls will fail.

**Comment by amanbha** [comment]
> > > @amanbha is this a dupe of something you opened in the past? My long-term memory is coming into play here..
> > 
> > 
> > @yaron2 It seems different, I think the issue which you are referring to is this one #1038
> 
> Yeah that's the one. The common theme is we should get the runtime to detect app restarts. This is useful for metrics too.

Yes, restarts should be handled.as per  #1038

**Comment by artursouza** [comment]
@amanbha @yaron2 (IMO) In a production environment, if daprd is proxying calls into my app, it should detect unhealthy instances and failover to a healthy instance - similar to a load balancer. Users should be aware that some instances are failing but their service should not experience reduced availability if healthy redundancy is available.

**Comment by yaron2** [comment]
> @amanbha @yaron2 (IMO) In a production environment, if daprd is proxying calls into my app, it should detect unhealthy instances and failover to a healthy instance - similar to a load balancer. Users should be aware that some instances are failing but their service should not experience reduced availability if healthy redundancy is available.

We have a mechanism for retries today that will retry on errors coming from the destination Dapr sidecar.

Essentially in Kubernetes, production environments are recommended to set up health probe on the container level. In that case, Kubernetes itself will take the Pod out of the load balancing rules and requests will be routed to healthy pods.

**Comment by artursouza** [comment]
@yaron2 Makes sense for K8s. What about for standalone mode?

**Comment by yaron2** [comment]
Standalone mode uses the components to resolve remote addresses, so this really comes down as a concern of the hosting platform.

You will get the same type of behavior in DC/OS, Swarm and hopefully Service Fabric.
For a local node setup, this is probably less useful.

**Comment by artursouza** [comment]
I was also thinking about making dapr cli exit when app exits. If we add the mode where dapr cli can run without app, we can add this feature to the cli as well.

**Comment by amanbha** [comment]
@artursouza @yaron2  Based on the discussion I think this is not a issue any more(definatley not P0) and can be closed in favor of the cli issue which @artursouza  explained above.

**Comment by yaron2** [comment]
> @artursouza @yaron2 Based on the discussion I think this is not a issue any more(definatley not P0) and can be closed in favor of the cli issue which @artursouza explained above.

I agree this is not P0. 
We should however talk about having actor sdks implement health check endpoints. If that turns into a separate issue I think we can close this one.

**Comment by artursouza** [comment]
This issue is originally to report the issue, so I will revert it back to describe the problem and not the solution. Then we can keep it open unless we decide that the current behavior is acceptable.

**Comment by LMWF** [comment]
> I was also thinking about making dapr cli exit when app exits. If we add the mode where dapr cli can run without app, we can add this feature to the cli as well.

Can you clarify how having the cli exit when the app does solves the problem?

**Comment by artursouza** [comment]
> > I was also thinking about making dapr cli exit when app exits. If we add the mode where dapr cli can run without app, we can add this feature to the cli as well.
> 
> Can you clarify how having the cli exit when the app does solves the problem?

It would only help in the context of cli - how this issue originated - but does not solve the root cause.

**Comment by yaron2** [comment]
## Design Proposal

Following is a design for implementing a health endpoint in the Actor runtime SDKs.

The Dapr runtime will continuously probe the health endpoint and upon 2 consecutive failures, 5 seconds apart, will remove itself from the placement ring.

Likewise when it receives an OK response after the same defined period, it will rejoin the ring.

The health endpoint url format:
```
<address:port>/healthz
```

A `200 OK` response is a guarantee for a successful liveness probe from the Dapr runtime side.
Any other response will count as a failure.

The `healthz` endpoint is named so to be consistent with Dapr's own health endpoint.

**Comment by amanbha** [comment]
> /cc @amanbha

lgtm

**Comment by yaron2** [comment]
This means we should have the endpoint inside the Dotnet SDK and Java SDK for this release.

cc @artursouza

**Comment by amanbha** [comment]
> This means we should have the endpoint inside the Dotnet SDK and Java SDK for this release.
> 
> cc @artursouza

Yes, and any third party actor implementations(outside of our SDKs0 should also have it, so we will have to document it and explain this scenario as well.

**Comment by yaron2** [comment]
If we have existing docs about the API endpoints used in an actor runtime, it should go in there.

**Comment by artursouza** [comment]
Shouldn’t this be an API beyond just actors?

**Comment by yaron2** [comment]
> Shouldn’t this be an API beyond just actors?

Probably not, as a failure here means for Dapr to rebalance the actors. For non actors, I'm not sure what a failure in the app would cause.

**Comment by artursouza** [comment]
Right now, it is only used for actors, but it can be used for other scenarios later. The API itself just says healthz, which is not specific to actors - which I agree with.

**Comment by yaron2** [comment]
> Right now, it is only used for actors, but it can be used for other scenarios later. The API itself just says healthz, which is not specific to actors - which I agree with.

Yeah. Let's focus on the immediate goal which is actors, then we can think next about other scenarios.

**Comment by artursouza** [comment]
I agree. I am just asking to keep the API outside the actor namespace.

**Comment by amanbha** [comment]
It mainly impacts actors, as no call would ever succeed for a particular actor id if user app is not responding. For non-actor calls the call is not tied to a specific instance.

**Comment by artursouza** [comment]
I agree. My point is that if we want to use this feature outside actors later, it will be a breaking change.

If we want to make this specific to Actors, we don’t need a new API. We can use the mandatory /Dapr/config as health check.

**Comment by amanbha** [comment]
> I agree. My point is that if we want to use this feature outside actors later, it will be a breaking change.
> 
> If we want to make this specific to Actors, we don’t need a new API. We can use the mandatory /Dapr/config as health check.

Yes, we can use config endpoint, my concern is with overloading because user code might be doing lots of processing on that route assuming its called only once, now we have to educate them that cache the values for subsequent calls as its also used as a health probe.

**Comment by artursouza** [comment]
Ok. So going back to the health API. I am proposing it to be outside the actor namespace. That is all.

**Comment by yaron2** [comment]
@artursouza are you talking about the actor namespace in the SDK? in that case, isn't it dependent on where the HTTP server sits? as its just a 200 OK response.

If the HTTP server is outside of the actors namespace/package, then its outside..

**Comment by artursouza** [comment]
Yes, but it depends on how we are extending the rest controller. So I can assume we agree that this is outside Actor namespace.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by amanbha** [comment]
I think you wanted to clarify if it will be a route like /actors/healthz or /healthz. It will be latter. For this change Dapr runtime will only use it for actors placement distribution.
